include/nfl/gmp.hpp

```diff
@@ -0,0 +1,232 @@
+#ifndef NFL_GMP_HPP
+#define NFL_GMP_HPP
+
+#include <vector>
+#include "nfl/poly.hpp"
+
+namespace nfl {
+
+template <class T, size_t Degree, size_t NbModuli>
+typename poly<T, Degree, NbModuli>::GMP poly<T, Degree, NbModuli>::gmp;
+
+/*
+ * Constructors
+ */
+template <class T, size_t Degree, size_t NbModuli>
+poly<T, Degree, NbModuli>::poly(mpz_t const& v) {
+  set_mpz(v);
+}
+
+template <class T, size_t Degree, size_t NbModuli>
+poly<T, Degree, NbModuli>::poly(mpz_class const& v) {
+  set_mpz({v}, (size_t)1);
+}
+
+template <class T, size_t Degree, size_t NbModuli>
+poly<T, Degree, NbModuli>::poly(std::initializer_list<mpz_class> const& values, size_t length) {
+  set_mpz(values, length);
+}
+
+template <class T, size_t Degree, size_t NbModuli>
+poly<T, Degree, NbModuli>::poly(mpz_t* const& values, size_t length) {
+  set_mpz(values, length);
+}
+
+template <class T, size_t Degree, size_t NbModuli>
+poly<T, Degree, NbModuli>::poly(mpz_class* const& values, size_t length) {
+  set_mpz(values, length);
+}
+
+/*
+ * set_mpz's functions
+ */
+
+template <class T, size_t Degree, size_t NbModuli>
+void poly<T, Degree, NbModuli>::set_mpz(mpz_t const& v) {
+  set_mpz({mpz_class(v)}, (size_t)1);
+}
+
+template <class T, size_t Degree, size_t NbModuli>
+void poly<T, Degree, NbModuli>::set_mpz(mpz_class const& v) {
+  set_mpz({v}, (size_t)1);
+}
+
+template <class T, size_t Degree, size_t NbModuli>
+void poly<T, Degree, NbModuli>::set_mpz(mpz_t *const &values, size_t length) {
+  std::vector<mpz_class> v(length);
+  for (size_t i = 0; i < length; i++) {
+    v[i] = mpz_class(values[i]);
+  }
+  set_mpz(v.begin(), v.end(), length);
+}
+
+template <class T, size_t Degree, size_t NbModuli>
+void poly<T, Degree, NbModuli>::set_mpz(mpz_class* const& values, size_t length) {
+  std::vector<mpz_class> v(length);
+  for (size_t i = 0; i < length; i++) {
+    v[i] = values[i];
+  }
+  set_mpz(v.begin(), v.end(), length);
```

Comment:
could this be just set_mpz(values, values + length) ? the length parameter looks redundant to me